File: clients/ios/AwareChat-iOS/AwareChat-iOS/Core/Services/MessagingClock.swift
import Foundation

protocol MessagingClock: Sendable {
  func now() -> Date
  func sleep(for duration: Duration) async throws
}

nonisolated struct SystemMessagingClock: MessagingClock {
  func now() -> Date { Date() }
  func sleep(for duration: Duration) async throws { try await Task.sleep(for: duration) }
}

nonisolated struct MessagingRetryPolicy: Sendable {
  static let acceptanceTimeout: Duration = .seconds(10)
  static let identificationTimeout: Duration = .seconds(10)
  private var index = 0
  private static let delays: [Duration] = [1, 2, 4, 8, 16, 30].map { .seconds($0) }

  mutating func nextDelay() -> Duration {
    let delay = Self.delays[index]
    index = min(index + 1, Self.delays.count - 1)
    return delay
  }

  mutating func reset() { index = 0 }
}

File: clients/android/AwareChat-Android/app/src/main/java/com/example/awarechat_android/core/service/MessagingClock.kt
package com.example.awarechat_android.core.service

import java.time.Instant
import kotlin.time.Duration
import kotlin.time.Duration.Companion.seconds
import kotlinx.coroutines.delay

interface MessagingClock {
    fun now(): Instant
    suspend fun sleep(duration: Duration)
}

class SystemMessagingClock : MessagingClock {
    override fun now(): Instant = Instant.now()

    override suspend fun sleep(duration: Duration) {
        delay(duration)
    }
}

internal class MessagingRetryPolicy {
    private var index = 0

    fun nextDelay(): Duration {
        val delay = delays[index]
        index = minOf(index + 1, delays.lastIndex)
        return delay
    }

    fun reset() {
        index = 0
    }

    companion object {
        val acceptanceTimeout = 10.seconds
        val identificationTimeout = 10.seconds
        private val delays = listOf(1, 2, 4, 8, 16, 30).map { it.seconds }
    }
}
